packages/protocol/test/identity/attestationsubsidy.ts

```diff
@@ -0,0 +1,307 @@
+    return authorizeFn(addr, signature.v, signature.r, signature.s, {
+      from: account,
+    })
+  }
+
+  beforeEach(async () => {
+    accountsInstance = await Accounts.new()
+    mockStableToken = await MockStableToken.new()
+    otherMockStableToken = await MockStableToken.new()
+    const mockValidators = await MockValidators.new()
+    attestations = await Attestations.new()
+    attestationSubsidy = await AttestationSubsidy.new()
+    random = await Random.new()
+    await random.initialize(256)
+    await random.addTestRandomness(0, '0x00')
+    mockLockedGold = await MockLockedGold.new()
+    registry = await Registry.new()
+    await accountsInstance.initialize(registry.address)
+    await registry.setAddressFor(CeloContractName.Validators, mockValidators.address)
+
+    await Promise.all(
+      accounts.map(async (account) => {
+        await accountsInstance.createAccount({ from: account })
+        await unlockAndAuthorizeKey(
+          KeyOffsets.VALIDATING_KEY_OFFSET,
+          accountsInstance.authorizeValidatorSigner,
+          account
+        )
+        await unlockAndAuthorizeKey(
+          KeyOffsets.ATTESTING_KEY_OFFSET,
+          accountsInstance.authorizeAttestationSigner,
+          account
+        )
+      })
+    )
+
+    mockElection = await MockElection.new()
+    await mockElection.setElectedValidators(
+      accounts.map((account) =>
+        privateKeyToAddress(getDerivedKey(KeyOffsets.VALIDATING_KEY_OFFSET, account))
+      )
+    )
+    await registry.setAddressFor(CeloContractName.Accounts, accountsInstance.address)
+    await registry.setAddressFor(CeloContractName.Random, random.address)
+    await registry.setAddressFor(CeloContractName.Election, mockElection.address)
+    await registry.setAddressFor(CeloContractName.LockedGold, mockLockedGold.address)
+    await attestations.initialize(
+      registry.address,
+      attestationExpiryBlocks,
+      selectIssuersWaitBlocks,
+      maxAttestations,
+      [mockStableToken.address, otherMockStableToken.address],
+      [attestationFee, attestationFee]
+    )
+
+    await attestationSubsidy.initialize(registry.address, {
+      from: subsidyOwner,
+    })
+
+    await attestations.__setValidators(
+      accounts.map((account) =>
+        privateKeyToAddress(getDerivedKey(KeyOffsets.VALIDATING_KEY_OFFSET, account))
+      )
+    )
+
+    callerMetaWallet = await MetaTransactionWallet.new()
+    await callerMetaWallet.initialize(caller)
+  })
+
+  describe('#initialize()', () => {
+    it('should not be callable again', async () => {
+      await assertRevert(
+        attestations.initialize(
+          registry.address,
+          attestationExpiryBlocks,
+          selectIssuersWaitBlocks,
+          maxAttestations,
+          [mockStableToken.address],
+          [attestationFee]
+        )
+      )
+    })
+  })
+
+  describe('#requestAttestationsWithSubsidy', () => {
```

Comment:
Would be also nice to see some negative cases with badly formed signature/inputs/etc.

Reply:
Yes I want to add more, tests are very much WIP.